File: 100Prisoners.cpp

/**
 * 100Prisoners.cpp
 *
 * Implementation of the loop strategy for the 100 prisoners problem (~31% win ratio).
 * usage: g++ 100Prisoners.cpp -o 100Prisoners && ./100Prisoners
 *
 * @author Maxamilian Demian
 * @link https://www.maxodev.org
 * @link https://github.com/Maxoplata/100-Prisoners
 */
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

int main() {
    int NUMBER_OF_TESTS = 10000; // the number of times to run the simulation
    int NUMBER_OF_PRISONERS = 100; // should be an even number for this problem

    // vars to keep track of our simulation wins and losses
    int totalWins = 0;
    int totalLosses = 0;

    // run each simulation
    for (int i = 0; i < NUMBER_OF_TESTS; i++) {
        // create the boxes
        std::vector<int> boxes(NUMBER_OF_PRISONERS);
        for (int boxNo = 1; boxNo <= NUMBER_OF_PRISONERS; boxNo++) {
            boxes[boxNo - 1] = boxNo;
        }

        // randomize the boxes
        unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
        shuffle (boxes.begin(), boxes.end(), std::default_random_engine(seed));

        // have the prisoners lost yet?
        bool didPrisonersLose = false;

        // each prisoner takes their chance at opening boxes
        for (int prisonerNumber = 1; prisonerNumber <= NUMBER_OF_PRISONERS; prisonerNumber++) {
            // prisoner has not found their number yet
            bool hasPrisonerFoundNumber = false;

            // prisoner will start by opening the box # matching their prisoner #
            int boxToOpen = prisonerNumber;

            // prisoner gets to open half of the boxes looking for their own number
            for (int openedBoxCount = 1; openedBoxCount <= (NUMBER_OF_PRISONERS / 2); openedBoxCount++) {
                // if the box they opened has their prisoner number, success! onto the next prisoner
                if (boxes[boxToOpen - 1] == prisonerNumber) { // - 1 because vector is zero-indexed
                    hasPrisonerFoundNumber = true;

                    break;
                }

                // otherwise, open the box # correlating to the # found in the box that was just opened
                boxToOpen = boxes[boxToOpen - 1];
            }

            // prisoner didn't find their number, everybody loses
            if (!hasPrisonerFoundNumber) {
                didPrisonersLose = true;

                break;
            }
        }

        if (didPrisonersLose) {
            totalLosses++;
        } else {
            totalWins++;
        }
    }

    printf("Wins/Losses: %d/%d (%f%% wins)\n", totalWins, totalLosses, (((float) totalWins / NUMBER_OF_TESTS) * 100));
}

File: 100Prisoners.rs
/**
 * 100Prisoners.rs
 *
 * Implementation of the loop strategy for the 100 prisoners problem (~31% win ratio).
 * usage: cargo build && ./100Prisoners
 *
 * @author Maxamilian Demian
 * @link https://www.maxodev.org
 * @link https://github.com/Maxoplata/100-Prisoners
 */
use rand::seq::SliceRandom;

const NUMBER_OF_TESTS: i32 = 10000; // the number of times to run the simulation
const NUMBER_OF_PRISONERS: i32 = 100; // should be an even number for this problem

fn main() {
	// vars to keep track of our simulation wins and losses
	let mut total_wins = 0;
	let mut total_losses = 0;

	// seed the rng
	let mut rng = rand::thread_rng();

	// run each simulation
	for _ in 0..NUMBER_OF_TESTS {
		// create the boxes
		let mut boxes: Vec<i32> = (1..(NUMBER_OF_PRISONERS + 1)).collect();

		// randomize the boxes
		boxes.shuffle(&mut rng);

		// have the prisoners lost yet?
		let mut did_prisoner_lose = false;

		// each prisoner takes their chance at opening boxes
		for prisoner_number in 1..(NUMBER_OF_PRISONERS + 1) {
			// prisoner has not found their number yet
			let mut has_prisoner_found_number = false;

			// prisoner will start by opening the box # matching their prisoner #
			let mut box_to_open = prisoner_number;

			// prisoner gets to open half of the boxes looking for their own number
			for _ in 0..(NUMBER_OF_PRISONERS / 2) {
				has_prisoner_found_number = false;

				// if the box they opened has their prisoner number, success! onto the next prisoner
				if boxes[(box_to_open - 1) as usize] == prisoner_number { // - 1 because array is zero-indexed
					has_prisoner_found_number = true;

					break;
				}

				// otherwise, open the box # correlating to the # found in the box that was just opened
				box_to_open = boxes[(box_to_open - 1) as usize];
			}

			// prisoner didn't find their number, everybody loses
			if !has_prisoner_found_number {
				did_prisoner_lose = true;

				break;
			}
		}

		if did_prisoner_lose {
			total_losses = total_losses + 1;
		} else {
			total_wins = total_wins + 1;
		}
	}

	println!("Wins/Losses: {}/{} ({}% wins)", total_wins, total_losses, (total_wins as f32 / NUMBER_OF_TESTS as f32) * 100.0);
}
